### ISSUE TITLE
Pipe commands change code in significant ways

### ISSUE THREAD

**Issue by mhanberg** [elixir-tools/next-ls#391]
## Descriptions

The pipe commands (in this case to-pipe), seem to be changing the formatting, and in some case, then semantics of code.

The following code changes a bracketless keyword list into a tuple with no list brackets, and switches the keyword style atom keys of a map into arrow style.

```elixir
NextLS.ASTHelpers.Env.build(ast |> Spitfire.parse!(:literal_encoder => &{:ok, {:__block__, &2, [&1]}}), %{
  line: 3,
  character: 4
})
```

becomes

```elixir
ast                # 👇 can see here the list is a tuple  👇 and👇 now have spaces  
|> Spitfire.parse!({:literal_encoder, &{:ok, {:__block__, & 2, [& 1]}}})
|> NextLS.ASTHelpers.Env.build(%{
  :line => 3, # 👈 is not a fat arrow style map
  :character => 4
})
```

cc @NJichev

**Comment by NJichev** [comment]
This is caused by the bad AST, since `:literal_encoder => ...` isn't valid syntax, if we want the keyword syntax there it has to be `literal_encoder: ...` or it needs the `%{}` literal if it's a map.
I think the only way we can ensure we don't mess with the original meaning of the code is if we work only with an `{:ok, ast}` from Spitfire? This would make the command less powerful though.

Here's how we parse the AST in the Pipe commands:
```
  defp parse(lines) do
    lines
    |> Enum.join("\n")
    |> Spitfire.parse(literal_encoder: &{:ok, {:__block__, &2, [&1]}})
    |> case do
      {:error, ast, _errors} ->
        {:ok, ast}

      other ->
        other
    end
  end
```
Basically this hits the error case from spitfire and we're "Ok, let's see what we can do".
In the valid case it does what we expect from what I saw, although it's a bit hard to reason where to put the cursor to make it do the correct thing.

Also this is probably a bug as well:
```
# cursor on pipe                    👇 
    NextLS.ASTHelpers.Env.build(ast |> Spitfire.parse!(%{:literal_encoder => &{:ok, {:__block__, &2, [&1]}}}), %{
    line: 3,
    character: 4
  })
```

Produces this:
```
    NextLS.ASTHelpers.Env.build(ast |> |>(Spitfire.parse!(%{:literal_encoder => &{:ok, {:__block__, & 2, [& 1]}}})), %{
    line: 3,
    character: 4
  })
```

**Comment by mhanberg** [comment]
Oh... I had a typo in the description. The original AST should have no errors. 😅

**Comment by mhanberg** [comment]
I'll re-confirm the repro later today.

**Comment by mhanberg** [comment]
https://github.com/elixir-tools/next-ls/assets/5523984/2291e489-4831-4707-94ce-72597ca13a09

So here i use `from-pipe` and then undo/re-do a couple times to highlight the problem. you can see it turns the keyword arg into a `:foo => value` style.

**Comment by mhanberg** [comment]
Might be a spitfire problem tho, I just confirmed that `code |> Code.string_to_quoted(token_metadata: true, columns: true, literal_encoder: &{:ok, {:__literal__, &2, [&1]}}) == code |> Spitfire.parse(literal_encoder: &{:ok, {:__literal__, &2, [&1]}})` outputs `false`

they should be identical.

they are without the literal encoder, so probably an issue there

**Comment by mhanberg** [comment]
alright, I patched spitfire, so you should be able to `mix deps.update spitfire` now (I also bumped the elxiir version, so make sure to run asdf/mise install)

i think there is a failing test now, but i'll leave that to you to fix @NJichev
